### ISSUE TITLE
Support using libxml SAX mode as the XML parser

### ISSUE THREAD

**Issue by mithro** [duck2/uxsdcxx#3]
As you showed previously, using a SAX parser has significant advantages in memory usage. Now that we are auto-generating the parser, it would be good to have the option between pugixml (for the highest speed) and a SAX parser for the lowest memory usage.

**Comment by duck2** [comment]
Using XPath with a SAX parser is not possible(or at least it kills its purpose because you build an in-memory representation), so this wouldn't work with anything under the [constraint components](https://www.w3.org/TR/xmlschema11-1/#Constraint_Summary) of XSD.

But it could be used to verify types, model groups, attributes, etc.

**Comment by mithro** [comment]
You should be possible to generate a post-structure creation checker?

**Comment by duck2** [comment]
Makes sense! If xml files can map to structs, maybe xpaths can somehow map to struct accesses.

**Comment by mithro** [comment]
I think that kind of makes sense...

**Comment by litghost** [comment]
I've done a quick test, and it looks like the pugixml parser isn't that much faster than libxml2's SAX mode, and the SAX mode uses significantly less memory.  Given the new interface based approached, could we implement a SAX mode?

**Comment by duck2** [comment]
That makes more sense since the interface is looking more like SAX.

Pugi is quite faster than libxml2 SAX, though. I have tested it a lot when doing [vtr-verilog-to-routing/pull/34](https://github.com/SymbiFlow/vtr-verilog-to-routing/pull/34), and however hard I tried to optimize the SAX-based parser(perfect hashes, indexed function lookups and all), it fell short of the unoptimized Pugi-based one.
